pfio/MultiGroupServer.F90

```diff
@@ -447,65 +447,103 @@ subroutine start_back(this, rc)
      type (c_ptr) :: address
      type (ForwardDataAndMessage), target :: f_d_m
      type (FileMetaData) :: fmd
-
+     logical :: flag
+     integer :: nwriter_per_node
+     integer, allocatable :: idleRank(:,:)   ! idle processors  
+     integer, allocatable :: num_idlePEs(:) ! how many idle processors in each node of backend server 
+     integer :: local_rank, node_rank, nth_writer
 
      call MPI_Comm_rank(this%back_comm, back_local_rank, ierr)
      thread_ptr => this%threads%at(1)
 
-     allocate(busy(this%nwriter-1), source =0)
     
      allocate(this%serverthread_done_msgs(1))
      this%serverthread_done_msgs(:) = .false.
- 
+     nwriter_per_node = this%nwriter/this%Node_Num
+     allocate(num_idlePEs(0:this%Node_Num-1))
+     allocate(idleRank(0:this%Node_Num-1, 0:nwriter_per_node-1), source=-1)
+
      if ( this%I_am_back_root ) then ! captain node is distributing work
+        ! all local ranks are idle at the beginning
+        do local_rank = 1, this%nwriter-1
+           node_rank = this%node_ranks(local_rank)
+           nth_writer = mod(local_rank, nwriter_per_node)
+           idleRank(node_rank, nth_writer) = local_rank
+        enddo    
+
+        num_idlePEs = count(idleRank /=-1, dim = 2)
+
         do while (.true.)
            ! 1) captain node of back_comm is waiting command from front_comm
            call MPI_recv( collection_id, 1, MPI_INTEGER, &
                 this%front_ranks(1), this%back_ranks(1), this%server_comm, &
                 MPI_STAT, ierr)
+
            if (collection_id >= 1) then ! front_comm is asking for a writing node 
-  
-               ! check idle woker
-               idle_worker = 0
-               do i = 1, this%nwriter -1
-                  if (busy(i) == 0) then
-                     idle_worker = i
-                     exit
-                  endif
-               enddo
-
-               ! if all workers are busy, wait for one
-               if (idle_worker == 0) then 
-                  call MPI_recv( idle_worker, 1, MPI_INTEGER, &
-                      MPI_ANY_SOURCE, stag, this%back_comm, &
-                      MPI_STAT, ierr)
-               endif
+
```

Comment:
This subroutine is getting long.   I encourage you to try to move some of this logic into well-named helper procedures to keep the logic in reasonably sized chunks.

Reply:
I have re-organized the codes. And it seems more clear now